### PR TITLE
chore(actors): refactor allgather

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Remove the actor-ness from allocation allgather.

## Test Plan
- [x] add units

## Commentary (optional)
EE tests, since only EE code uses this https://github.com/determined-ai/determined-ee/pull/910

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
DET-9611

### CODE REVIEW

File: master/internal/task/allgather.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/determined-ai/determined/master/internal/task/allgather"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AllGather blocks until `numPeers` with the same `allocationID` are waiting and then returns the
+// data from all those peers. It returns an error in the if the call returns early, without data,
+// for any reason. Only one call may connect per `id`.
+func AllGather(
```
Comment (stoksc) [inline master/internal/task/allgather.go:16]: i'll probably ticket making `allgather.Join` sync just like this, removing `allgather.Leave`, and probably removing a lot of the code (watchers in particular). but in the spirit of keeping the changes super minimal, i'm gonna keep most of the weirdness

File: master/internal/task/allgather.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/google/uuid"
+
+	"github.com/determined-ai/determined/master/internal/task/allgather"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AllGather blocks until `numPeers` with the same `allocationID` are waiting and then returns the
+// data from all those peers. It returns an error in the if the call returns early, without data,
```
Comment (erikwilson) [inline master/internal/task/allgather.go:14]: nit: comment needs some cleanup

**Review by erikwilson — APPROVED** [review]

File: master/internal/task/allgather/allgather.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	return w
+}
+
+func (g *allGather) leave(id uuid.UUID) (empty bool) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	c, ok := g.watchers[id]
+	if !ok {
+		return
```
Comment (maxrussell) [inline master/internal/task/allgather/allgather.go:109]: Maybe I'm missing something, but seems strange to say the watchers isn't empty whenever we don't find the watcher they pass in the id of.

**Review by maxrussell — APPROVED** [review]
LGTM as long as you think that `leave` thing is correct.

File: master/internal/task/allgather/allgather.go (hunk 3 above)
Comment (stoksc) [reply master/internal/task/allgather/allgather.go:109]: you are totally right! curse named returns

File: master/internal/task/preemptible/preemptible.go [hunk 4]
```diff
@@ -20,7 +20,7 @@ var ErrPreemptionDisabled = fmt.Errorf("allocation is not preemptible")
 // DefaultTimeout is the delay before the deadline exceeded callback passed to preempt is called.
 var DefaultTimeout = time.Hour
 
-// Watcher contains a channel which can be polled for a preemption signal.
+// Watcher signals all gather completion via a channel which is closed upon said completion.
```
Comment (stoksc) [inline master/internal/task/preemptible/preemptible.go:23]:
```suggestion
// Watcher contains a channel which can be polled for a preemption signal.
```

**Pull request merged**